primitives/arithmetic/src/per_things.rs

```diff
@@ -30,29 +30,55 @@ pub trait PerThing: Sized + Saturating + Copy {
 	/// The data type used to build this per-thingy.
 	type Inner: BaseArithmetic + Copy;
 
-	/// accuracy of this type
-	const ACCURACY: Self::Inner;
+	/// A data type larger than `Self::Inner`, used to avoid overflow in some computations.
+	type Upper: BaseArithmetic + Copy + From<Self::Inner> + TryInto<Self::Inner>;
 
-	/// NoThing
-	fn zero() -> Self;
+	/// The accuracy of this type.
+	const ACCURACY: Self::Inner;
 
-	/// `true` if this is nothing.
-	fn is_zero(&self) -> bool;
+	/// Equivalent to `Self::from_parts(0)`.
+	fn zero() -> Self { Self::from_parts(Self::Inner::zero()) }
 
-	/// Everything.
-	fn one() -> Self;
+	/// Return `true` if this is nothing.
+	fn is_zero(&self) -> bool { self.deconstruct() == Self::Inner::zero() }
 
-	/// Consume self and deconstruct into a raw numeric type.
-	fn deconstruct(self) -> Self::Inner;
+	/// Equivalent to `Self::from_parts(Self::ACCURACY)`.
+	fn one() -> Self { Self::from_parts(Self::ACCURACY) }
 
-	/// From an explicitly defined number of parts per maximum of the type.
-	fn from_parts(parts: Self::Inner) -> Self;
+	/// Return `true` if this is one.
+	fn is_one(&self) -> bool { self.deconstruct() == Self::ACCURACY }
 
-	/// Converts a percent into `Self`. Equal to `x / 100`.
-	fn from_percent(x: Self::Inner) -> Self;
+	/// Build this type from a percent. Equivalent to `Self::from_parts(x * Self::ACCURACY / 100)`.
+	fn from_percent(x: Self::Inner) -> Self {
+		Self::from_parts(x.min(100.into()) * (Self::ACCURACY / 100.into()))
+	}
 
 	/// Return the product of multiplication of this value by itself.
-	fn square(self) -> Self;
+	fn square(self) -> Self {
+		let p = Self::Upper::from(self.deconstruct());
+		let q = Self::Upper::from(Self::ACCURACY); 
+		Self::from_rational_approximation(p * p, q * q)
+	}
+
+	/// Saturating reciprocal multiplication. Compute `x / self`, saturating at the numeric 
+	/// bounds instead of overflowing.
+	fn saturating_reciprocal_mul<N>(self, x: N) -> N
+	where N: Into<Self::Upper> + UniqueSaturatedFrom<Self::Upper> {
```

Comment:
note that the bound `Into<Self::Upper>` is a strong constrain, for example usual arithmetic used in substrate are just implementing the trait AtLeast32Bit, and are generally more than u32.

On the other side Perbill or other per_things are usually using u64 as Upper type. so this is limiting the usage.

Reply:
Changed to `UniqueSaturatedInto<Self::Inner>`. Better you think?

Reply:
I don't think so, this saturation is very likely to happen for lots of value I guess.
for instance a currency amount can likely not fit into u64.

Maybe there is a way to solve it similarly to `fn mul` or maybe it is better to actually compute everything into the arithmetic space of `N`

Reply:
I'm ok to merge with `Into<Self::Inner>` just I feel like maybe we could improve this

Reply:
also I'm pretty we can make a similar implementation as `fn mul`

Reply:
> I don't think so, this saturation is very likely to happen for lots of value I guess.
> for instance a currency amount can likely not fit into u64.

I don't think this is true. If you are handling values that regularly saturate `Inner` you should choose a bigger `Inner`, or implement one.

Reply:
> also I'm pretty we can make a similar implementation as `fn mul`

What aspects of the `mul` implementation would improve this one?  The `mul` rounds to the nearest part, whereas these ones are meant to round down.

Reply:
I could add various other operations, (eg. checked as well as saturating), if that would be useful, but I don't know if anyone needs them so could probably go in a later PR.

Reply:
> > I don't think so, this saturation is very likely to happen for lots of value I guess.
> > for instance a currency amount can likely not fit into u64.
> 
> I don't think this is true. If you are handling values that regularly saturate `Inner` you should choose a bigger `Inner`, or implement one.

Perbill is used a lot in the code base with Currency and in kusama currency is u128 and usual amount does overflow u32 (`u32== Perbill::Inner`). And I don't think this is a wrong usage

Reply:
> > also I'm pretty we can make a similar implementation as `fn mul`
> 
> What aspects of the `mul` implementation would improve this one? The `mul` rounds to the nearest part, whereas these ones are meant to round down.

implementation of `Mul` doesn't saturate the input `x: N` to u32 it computes an accurate multiplication (with rounding stuff) and doesn't saturate (as Per_things are for now in [0, 1]).

if I use some value in u128 and want to multiply with some Perbill the implementation Mul give a correct result whereas this implementation is very inaccurate if the input is more than `2^32`

Reply:
I've changed `saturating_reciprocal_mul` to use the same remainder calculation as `mul` so it only saturates `N`.  Great suggestion, thanks